services/kvstore/src/main/java/tech/pegasys/pantheon/services/kvstore/ColumnarRocksDbKeyValueStorage.java

```diff
@@ -0,0 +1,326 @@
+              .setMaxOpenFiles(rocksDbConfiguration.getMaxOpenFiles())
+              .setMaxBackgroundCompactions(rocksDbConfiguration.getMaxBackgroundCompactions())
+              .setStatistics(stats)
+              .setCreateMissingColumnFamilies(true)
+              .setEnv(
+                  Env.getDefault()
+                      .setBackgroundThreads(rocksDbConfiguration.getBackgroundThreadCount()));
+
+      txOptions = new TransactionDBOptions();
+      final List<ColumnFamilyHandle> columnHandles = new ArrayList<>(columnDescriptors.size());
+      db =
+          TransactionDB.open(
+              options,
+              txOptions,
+              rocksDbConfiguration.getDatabaseDir().toString(),
+              columnDescriptors,
+              columnHandles);
+
+      final Map<BytesValue, String> segmentsById =
+          segments.stream()
+              .collect(
+                  Collectors.toMap(segment -> BytesValue.wrap(segment.getId()), Segment::getName));
+
+      final ImmutableMap.Builder<String, ColumnFamilyHandle> builder = ImmutableMap.builder();
+      for (final ColumnFamilyHandle columnHandle : columnHandles) {
+        final String segmentName = segmentsById.get(BytesValue.wrap(columnHandle.getName()));
+        if (segmentName != null) {
+          builder.put(segmentName, columnHandle);
+        } else {
+          builder.put(DEFAULT_COLUMN, columnHandle);
+        }
+      }
+      columnHandlesByName = builder.build();
+
+      readLatency =
+          metricsSystem
+              .createLabelledTimer(
+                  PantheonMetricCategory.KVSTORE_ROCKSDB,
+                  "read_latency_seconds",
+                  "Latency for read from RocksDB.",
+                  "database")
+              .labels(rocksDbConfiguration.getLabel());
+      removeLatency =
+          metricsSystem
+              .createLabelledTimer(
+                  PantheonMetricCategory.KVSTORE_ROCKSDB,
+                  "remove_latency_seconds",
+                  "Latency of remove requests from RocksDB.",
+                  "database")
+              .labels(rocksDbConfiguration.getLabel());
+      writeLatency =
+          metricsSystem
+              .createLabelledTimer(
+                  PantheonMetricCategory.KVSTORE_ROCKSDB,
+                  "write_latency_seconds",
+                  "Latency for write to RocksDB.",
+                  "database")
+              .labels(rocksDbConfiguration.getLabel());
+      commitLatency =
+          metricsSystem
+              .createLabelledTimer(
+                  PantheonMetricCategory.KVSTORE_ROCKSDB,
+                  "commit_latency_seconds",
+                  "Latency for commits to RocksDB.",
+                  "database")
+              .labels(rocksDbConfiguration.getLabel());
+
+      if (metricsSystem instanceof PrometheusMetricsSystem) {
+        RocksDBStats.registerRocksDBMetrics(stats, (PrometheusMetricsSystem) metricsSystem);
```

Comment:
Do the rocks db stats need to be made aware of the column data store?  i.e. will the keys we fish for match up?

Reply:
I don't know but there doesn't seem to be a mention of columns here: https://github.com/facebook/rocksdb/wiki/Statistics. Is there an example statistic you had in mind?

Reply:
I was thinking out loud to see if stuff like caching was done per column.  If the docs don't mention it then I don't think it is done.